docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,229 @@
+
+To solve a problem, we'll have a global private mapping from `TenantId` to `TenantAccessor`, accessible only by `with_tenant`. `TenantAccessor` is a kind of guard similar to `Arc` in Rust but not the same. It allows running the closures with the tenant in scope but won't run them if it's not in the `Active` state, and this accessor is responsible for all cancellations.
+
+The interface:
+
+```rust
+impl TenantAccessor {
+    /// Creates a new accessor. Check the state is not `TenantState::Infant`.
+    /// Spawns a closure in tokio runtime to wait for the state to become dead.
+    /// Then, the tenant is shut down.
+    pub fn new(tenant: Tenant) -> Self;
+
+    /// Subscribe for the tenant drop. If the tenant is already dropped, then
+    /// watcher will also reflect that.
+    pub fn subscribe_for_shutdown(&self) -> DropWatcher;
+
+    /// Applies the specified function to the tenant. Returns error if the tenant is
+    /// not ready to launch the closure.
+    pub fn with_tenant<F, O>(&self, func: F) -> anyhow::Result<O>
+    where
+    F: FnOnce(TenantSyncRef) -> O {}
+
+    /// Applies the specified function to the tenant. Returns error if the tenant is
+    /// not ready to launch the closure.
+    pub async fn with_tenant_async<F, T, O>(&self, func: F) -> anyhow::Result<O>
+    where
+    F: FnOnce(TenantAsyncRef) -> T,
+    T: Future<Output = O> {}
+}
+
+impl Drop for TenantAccessor {
+    // schedule a drop of itself in tokio runtime with necessary joins...
+}
+```
+
+### `TenantAccessor` and states
+
+This is how the accessor behaves when it sees different states:
+
+1. `Infant` - we cannot create a guard around the tenant at this stage. We should start loading some layers and become `Loading` or `Downloading` first and only after creating an accessor.
+2. `Loading` - we cannot run any operations on the tenant until it is either in an `Active` or `Broken` state.
+3. `Broken` - we cannot run any operations on the tenant.
+4. `Downloading` - we cannot run any operations on the tenant until it is either in an `Active` or `Broken` state.
+5. `Active` - the tenant can run any closures. **That means it's the only way to run any closures on tenants**.
+6. `ShuttingDown` - we cannot run any operations on the tenant.
+7. `Shutdown` - we cannot run any operations on the tenant.
+
+### `TenantAccessor` and cancelations
+
+One great question is: what guarantees do we have about tenants inside the closures we're running?
+
+After we introduced the `with_tenant` functions, there are a bunch of closures running. The only way to run a closure is when we are under an `Active` state. What if one closure wants to run under this state, and another decides to shut down the tenant? Do we want to continue executing?
+
+In this specific RFC:
+
+1. We use a `StateKeeper` / `StateSubscriber` / `StateHider` primitive to ensure all closures are gone, and running closures can't tell the difference.
```

Comment:
Sorry, I haven't deleted this when cleaning up. Fixed.

Transitions are done inside the closure. But the core thing we're using is that it's only possible to have an `Active` state inside when you start running the closure and `ShuttingDown` if someone decides to kill the tenant. Since we only have 1 possible state change here, it's straightforward to manage. I think that could benefit from an example. Could you find something that's looking not transparent from the RFC?

Reply:
[Here's the discussion link](https://github.com/neondatabase/neon/pull/2899#discussion_r1031428734).